File: frontend/test/utils/writing_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:srbski_read/utils/writing.dart';

/// Правила отбора слов обязаны совпадать с `web/src/lib/writing.ts`: карточки
/// синхронизируются, и слово, доступное для письма на телефоне, должно быть
/// доступно и в браузере. Примеры здесь намеренно те же.
void main() {
  group('повторение письмом', () {
    test('берёт отдельные слова', () {
      for (final word in ['kuća', 'књига', 'razumeti', 'ђак']) {
        expect(writable(word), isTrue, reason: word);
      }
    });

    // Фразы отсеиваются намеренно: писать рукой предложение долго, и
    // вспоминается оно не так, как слово.
    test('не берёт фразы', () {
      for (final phrase in [
        'dobar dan',
        'како сте',
        'ja sam student',
        '  ',
        '',
      ]) {
        expect(writable(phrase), isFalse, reason: phrase);
      }
    });

    test('не берёт слишком короткое и слишком длинное', () {
      expect(writable('a'), isFalse);
      expect(writable('a' * 25), isFalse);
      expect(writable('a' * 24), isTrue);
    });

    // Сербская диакритика — одна кодовая точка, но обычный split по индексу
    // разрубил бы суррогатную пару.
    test('делит слово по видимым буквам', () {
      expect(wordLetters('kuća'), ['k', 'u', 'ć', 'a']);
      expect(wordLetters('ђак'), ['ђ', 'а', 'к']);
      expect(wordLetters('čas'), ['č', 'a', 's']);
    });
  });
}

File: web/src/lib/writing.test.ts
import { describe, expect, it } from 'vitest';

import { letters, writable } from './writing';

describe('повторение письмом', () => {
  it('берёт отдельные слова', () => {
    for (const word of ['kuća', 'књига', 'razumeti', 'ђак']) {
      expect(writable(word)).toBe(true);
    }
  });

  // Фразы отсеиваются намеренно: писать рукой предложение долго, и
  // вспоминается оно не так, как слово.
  it('не берёт фразы', () => {
    for (const phrase of [
      'dobar dan',
      'како сте',
      'ja sam student',
      '  ',
      '',
    ]) {
      expect(writable(phrase)).toBe(false);
    }
  });

  it('не берёт слишком короткое и слишком длинное', () => {
    expect(writable('a')).toBe(false);
    expect(writable('a'.repeat(25))).toBe(false);
    expect(writable('a'.repeat(24))).toBe(true);
  });

  // Диакритика в сербской латинице — часть буквы. Разбиение по кодовым
  // единицам дало бы «пустые» буквы, которые не нарисовать.
  it('делит слово по видимым буквам', () => {
    expect(letters('kuća')).toEqual(['k', 'u', 'ć', 'a']);
    expect(letters('ђак')).toEqual(['ђ', 'а', 'к']);
    // Составная запись «c» + комбинирующая гачек — та же буква «č».
    expect(letters('čas')).toEqual(['č', 'a', 's']);
  });
});
